rfcs/20240325-reorganize-gpu/README.md

```diff
@@ -0,0 +1,117 @@
+clustered around vendors therefore the new schema should not cause a lot of
+confusion among the developers, it should also positively affect the
+implementation cost
+* The schema also provides sufficient configurability, e.g
+the generic vendor can be enabled along with nvidia or amd, or it can be
+enabled individually
+
+```bash
+├── cpu/ # CPU code
+├── sycl/ # A common place for basic CPU and GPU SYCL abstractions, e.g. `sycl_cpu_engine_t`, `sycl_gpu_engine_t`, etc.
+└── gpu/ # GPU code. Basic GPU code resides in the GPU directory directly, e.g. `gpu_engine_t`.
+    ├── intel/ # Everything that is related to Intel
+    │   ├── compute/ # Compute layer abstractions
+    │   ├── ocl/ # OpenCL kernels and abstractions (very similar to the current state of `gpu/ocl` directory)
+    │   ├── jit/ # JIT kernels/generators and abstractions (very similar to the current state of `gpu/jit` directory)
+    │   ├── sycl/ # Intel related SYCL functionality/abstractions (e.g. SYCL kernels that use eSIMD, Intel specific extensions, etc)
+    │   └── ...
+    ├── nvidia/ # Everything that is related to NVIDIA
+    │   ├── sycl # NVIDIA related functionality/abstractions (cuDNN/cuBLAS based and CUDA specific SYCL kernels)
+    │   └── ...
+    ├── amd/ # Everything that is related to AMD
+    │   ├── sycl/ # AMD related functionality/abstractions (MIOpen/rocBLAS based and HIP specific SYCL kernels)
+    │   └── ...
+    └── generic/ # Everything that is related to generic kernels
+        └── sycl/ # generic SYCL kernels and related abstractions
+
+```
+
+### Affected Basic Abstractions
+
+The new schema will require moving a lot of parts of the library around
+and while most of the changes are probably just an implementation detail
+there are a few major changes that have to be described in this RFC.
+
+#### Engine
+
+Engine is the most heavily affected abstraction by the changes.
+There is currently no `gpu_engine_t` class as `compute_engine_t` is used as
+an Intel centric alternative of it. In order to have a common basic
+abstraction at the GPU level the `gpu_engine_t` class will be introduced.
+
+The current inheritance chains for SYCL and OpenCL GPU engines are
+the following:
+* SYCL Intel and generic: `engine_t` -> `compute_engine_t` -> `sycl_engine_base_t` -> `sycl_gpu_engine_t`
+* SYCL NVIDIA: `engine_t` -> `compute_engine_t` -> `sycl_engine_base_t` -> `sycl_cuda_engine_t`
+* SYCL AMD: `engine_t` -> `compute_engine_t` -> `sycl_engine_base_t` -> `sycl_hip_engine_t`
+* OpenCL (only Intel): `engine_t` -> `compute_engine_t` -> `ocl_gpu_engine_t`
+
+With the new schema they will be changed to:
+* SYCL Intel and generic: `engine_t` -> `gpu_engine_t` -> (`gpu_intel_engine_t` -> `compute_engine_t`) -> `sycl_engine_base_t` -> `sycl_gpu_intel_engine_t`
+* SYCL NVIDIA: `engine_t` -> `gpu_engine_t` -> (`gpu_intel_engine_t` -> `compute_engine_t`) -> `sycl_engine_base_t` -> `sycl_gpu_nvidia_engine_t`
```

Comment:
Though it's orthogonal to the RFC I feel like this might be a chance to drop prefixes in all related classes: e.g. `dnnl::impl::gpu::nvidia::sycl_gpu_nvidia_engine_t` -> `dnnl::impl::gpu::nvidia::engine_t`.

Reply:
We can definitely do that however I'm concerned about possible clashing problems like the one we have with two `sycl` namespaces.

In your example you skipped that `sycl` namespace, was it on purpose? Do you suggest to assume that everything that is nvidia/amd related is sycl related? If so, it looks a bit confusing to me.

Reply:
Well, if Nvidia/AMD implementations are all SYCL-based than having an extra directory is not that useful (maybe only for consistency).

Another observation is that current `src/gpu/ocl` combines OpenCL-specific functionality (opposite to SYCL which is split between `src/sycl` and `src/gpu/sycl`) and "pseudo"-OpenCL implementations which are in fact used with both OpenCL/SYCL (moreover there are fully OpenCL-agnostic implementations in this directory, e.g. [this](https://github.com/oneapi-src/oneDNN/blob/main/src/gpu/ocl/gemm_inner_product.hpp)). So if you have any ideas how to make it more consistent - this would be nice.

Reply:
> Well, if Nvidia/AMD implementations are all SYCL-based than having an extra directory is not that useful (maybe only for consistency).

The extra directory doesn't make much sense indeed because it doesn't look like we'll need anything other than SYCL to support Nvidia and AMD. On the other hand, we will have `sycl` namespace for intel vendor (`gpu::intel::sycl`). Anyway, in the bigger scheme of things it's not going to make a big difference so I'm fine with either way.

>  So if you have any ideas how to make it more consistent - this would be nice.

* I don't think we could get rid of `src/sycl` because we have to support CPU and GPU engines and that is the place for common SYCL code. If we want to get rid of it then we will have to move the common code to `src/common/sycl` and the CPU code to `src/cpu/sycl`. I don't know if we really need it because it doesn't give us much.
* We have a bunch of runtime agnostic primitives. They are basically comprised of nested primitives (e.g. sum, concat, gemm). We can put them directly in `src/gpu/generic` without any subdirectories.